internal/exec/util/file.go

```diff
@@ -151,25 +151,27 @@ func (u Util) WriteLink(s types.Link) error {
 }
 
 func (u Util) SetPermissions(mode *int, node types.Node) error {
-	if mode != nil {
-		if err := os.Chmod(node.Path, toFileMode(*mode)); err != nil {
-			return fmt.Errorf("failed to change mode of %s: %v", node.Path, err)
-		}
-	}
-
+	// Set ownership and then permissions, https://man7.org/linux/man-pages/man2/lchown.2.html
+	// "... When the owner or group of an executable file is changed by an
+	// unprivileged user, the S_ISUID and S_ISGID mode bits are cleared..."
 	defaultUid, defaultGid, _ := getFileOwnerAndMode(node.Path)
 	uid, gid, err := u.ResolveNodeUidAndGid(node, defaultUid, defaultGid)
-	if err != nil {
-		return fmt.Errorf("failed to determine correct uid and gid for %s: %v", node.Path, err)
-	}
 	if err := os.Lchown(node.Path, uid, gid); err != nil {
 		return fmt.Errorf("failed to change ownership of %s: %v", node.Path, err)
 	}
+	if mode != nil {
+		if err != nil {
+			return fmt.Errorf("failed to determine correct uid and gid for %s: %v", node.Path, err)
+		}
```

Comment:
I think that should be moved earlier above.

Reply:
So sorry, do you mean outside of the function its in? I believe its already at the highest point in the set permissions function it can be.